packages/inter-protocol/src/auction/auctionBook.js

```diff
@@ -0,0 +1,319 @@
+    if (isActive(auctionState) && ratioGTE(price, curAuctionPrice)) {
+      collateralSold = settle(seat, want);
+
+      if (AmountMath.isEmpty(seat.getCurrentAllocation().Currency)) {
+        seat.exit();
+        return;
+      }
+    }
+
+    const stillWant = AmountMath.subtract(want, collateralSold);
+    if (!AmountMath.isEmpty(stillWant)) {
+      priceBook.add(seat, price, stillWant, timestamp);
+    } else {
+      seat.exit();
+    }
+  };
+
+  const acceptDiscountOffer = (
+    seat,
+    discount,
+    want,
+    timestamp,
+    auctionState,
+  ) => {
+    trace('accept discount');
+    let collateralSold = AmountMath.makeEmptyFromAmount(want);
+
+    if (
+      isActive(auctionState) &&
+      isDiscountedPriceHigher(discount, curAuctionPrice, lockedPrice)
+    ) {
+      collateralSold = settle(seat, want);
+      if (AmountMath.isEmpty(seat.getCurrentAllocation().Currency)) {
+        seat.exit();
+        return;
+      }
+    }
+
+    const stillWant = AmountMath.subtract(want, collateralSold);
+    if (!AmountMath.isEmpty(stillWant)) {
+      discountBook.add(seat, discount, stillWant, timestamp);
+    } else {
+      seat.exit();
+    }
+  };
+
+  return Far('AuctionBook', {
+    addAssets(assetAmount, sourceSeat) {
+      trace('add assets');
+      assetsForSale = AmountMath.add(assetsForSale, assetAmount);
+      atomicRearrange(
+        zcf,
+        harden([[sourceSeat, collateralSeat, { Collateral: assetAmount }]]),
+      );
+    },
+    settleAtNewRate(reduction) {
+      curAuctionPrice = multiplyRatios(reduction, lockedPrice);
+
+      const pricedOffers = priceBook.offersAbove(curAuctionPrice);
+      const discOffers = discountBook.offersAbove(reduction);
+
+      // requested price or discount gives no priority beyond specifying which
+      // round the order will be service in.
+      const prioritizedOffers = [...pricedOffers, ...discOffers].sort(
+        ([a], [b]) => TimeMath.compareAbs(keyToTime(a), keyToTime(b)),
+      );
+
+      trace(`settling`, pricedOffers.length, discOffers.length);
+      prioritizedOffers.forEach(([key, { seat, price: p, wanted }]) => {
+        const collateralSold = settle(seat, wanted);
+
+        if (AmountMath.isEmpty(seat.getCurrentAllocation().Currency)) {
+          seat.exit();
+          if (p) {
+            priceBook.delete(key);
+          } else {
+            discountBook.delete(key);
+          }
+        } else if (!AmountMath.isEmpty(collateralSold)) {
+          if (p) {
+            priceBook.updateReceived(key, collateralSold);
+          } else {
+            discountBook.updateReceived(key, collateralSold);
+          }
+        }
+      });
+    },
+    getCurrentPrice() {
+      return curAuctionPrice;
+    },
+    hasOrders() {
+      return discountBook.hasOrders() || priceBook.hasOrders();
+    },
+    lockOraclePriceForRound() {
+      trace(`locking `, updatingOracleQuote);
+      lockedPrice = updatingOracleQuote;
```

Comment:
there are lots of prices in scope. consider something more descriptive,
```suggestion
      priceQuoteAtStart = latestOracleQuote;
```

Reply:
The more important aspect of this value is that it's locked before the round starts and (intended to be) visible to vault owners so they know how to protect themselves.  

I've already talked about `updating`  vs. `latest`.